README.md
# react-calendarien
---
리액트 달력입니다.<br/>
아직 완벽하진 않습니다.<br/>
현재 계속해서 개발중이며, 최적화 진행중입니다.<br/>

!! **주의** IE 에서는 아직 동작하지 않습니다. (Edge 는 가능합니다.)

## Props
name|type|default|description
----|----|-------|-----------
style|object|{ <br/>width:'100%'<br/>height: 'inherit'<br/>color: '#262626'<br/>}|layout 의 스타일을 담을 수 있습니다.
getValue|function||선택한 날의 정보를 가져올 수 있습니다. <br/> default return "yyyy mm dd"
customizeIcon|array|[ ]|버튼을 커스터마이징 할 수 있습니다. <br/> default icon <, >
layoutOption|array|[ ]|layout 의 디자인을 조금 변경가능하다. <br/> 값 .. ['outline', 'radius']
setFormat|string|yyyy mm dd|getValue 에서 return 되는 데이터의 형태를 바꿀 수 있습니다.
setDate|string||원하는 날의 달력을 render 시에 보여준다.<br/>하단에 'MY DATE' 버튼이 함께 생성된다.
visibleToday|boolean|false|하단 "TODAY" 버튼 가시화
visibleMyDate|boolean|false|setDate 옵션 추가시 자동적으로 나오는 'MY DATE' 버튼의 가시화를 정한다.
disabled|boolean|false|이전 날짜는 선택하지 못하도록 막는 기능이다.
theme|string||달력테마 디자인을 설정합니다.<br/> darkblue, forest, gd_darkblue, gd_sunset<br/> * gd가 붙은건 그라데이션을 의미함
mode|string||모드를 설정 하여 다양한 형태의 달력을 출력한다. <br/> input > input 에 focus 를 주면 달력이 나온다.

## Useage
create-react-app 을 사용했을 때의 경우 아래와 같이 테스트 합니다.<br/>


### 1. default example
---
```jsx
import React, { Component } from "react";
import Calendarien from './Calendarien';

class App extends Component {
    render() {
        return (
            <div>
                <Calendarien props... />
            </div>
        )
    }
} 
```

### 2. real time change 'setDate' example
---
```jsx
class App extends Component {

    handleDate = ({ value : date }) => {
         this.setState({ date });
    }

    state = {
        date: ''
    }
    
    render () {
        const { date } = this.state;
        return (
            <div>
                <Calendarien setDate={date} />
                <input 
                    type="text"  
                    onKeyPress={({key, target}) => key === 'Enter' && this.handleDate(target)} 
                />
            </div>
        )
    }
}
```
> date form 이 맞지 않을시 console.warn 을 출력합니다.

## Will be Update Function or Value
- mode [ <del><i>input</i></del>, <i>modal</i> ]
- allowRange
- View for Reactive Type ( mobile, web, tablet )

## If you are using an IE browser, you should read this
!! IE 적용시 react-app-polyfill 을 추가해야합니다.<br/>

<code>
> yarn add react-app-polyfill
</code>

```jsx
// index.js
import 'react-app-polyfill/ie9';
import 'react-app-polyfill/ie11';
(... import code)
(... render code)
```

## Connect
Made by demnodey
- blog :: https://blog.naver.com/psj9102
- email :: psj9102@naver.com

## License
react-calendarien is released under the MIT license

HISTORY.md
# HISTORY

### Will Update Function ::

- allowRange 기능 추가하기

## Update #10
#### 2018 12 05
- 디자인 부분수정
- mode 시 'TODAY' 버튼 기능 개선 ( today 클릭시 해당 날짜가 입력된다. )
---

## Update #9
#### 2018 12 03
- input mode 추가
- Button 및 기타 전체적인 디자인 수정
- 'TODAY' 버튼 기능 개선 ( 오늘 날짜 선택됨 )
- next, prev 기본 아이콘을 html unicode icon 으로 교체 
---

## Update #8
#### 2018 11 29
- 코드 최소화
- PropTypes 적용 안되있는 props 적용
- IE 브라우저 문제점 해결. ( development 모드에서 cdn polyfill script 링크가 필요하다. ) 
  - 1.index.html에 script 태그 적용 ( 링크 :: //cdn.polyfill.io/v2/polyfill.min.js )
  - 2.ployfill 적용하기 > yarn add react-app-polyfill ( import 'react-app-polyfill/ie9 or ie11')
- IE, Edge 에서 나던 Date 연산 오류 수정
- Calendarien 폴더 위치 수정
---

## Update #7
#### 2018 11 27
- 버그 수정
- 토이 프로젝트에 적용해 본 후 디자인 및 기능에 대한 문제점 수정필요 ( 진행중... )
- theme 추가 ( darkblue, forest, gd_darkblue, gd_sunset ) > Layout.css 에서 수정가능
- setDate 실시간 수정 가능하게 해놨음
---

## Update #6
#### 2018 11 26
- 몇가지 버그 수정 및 layout 디자인 옵션추가. outline, radius
- DayOfWeek 추가
- disabled 기능 추가 - 코드의 복잡함으로 인한 코드수정 예정 ( 진행중 ... )
---

## Update #5
#### 2018 11 23
- 이전달, 다음달 랜더링 완료
- 18/11/25 git 최신화
- 코드가 복잡해짐
---

## Update #4
#### 2018 11 22
- Viewer(Body)에 관련된 기본적인 기능 80% 정도 완료
- git 최신화는 하지 않았음.
- 파일구조가 복잡하고, 코드의 가독성이 떨어진다. 기능이 많이 부족하다.
---

## Update #3
#### 2018 11 22
- Header, Viewer 기본적인 달력기능 70%정도 완료 후, 본격적 README.md 작성 및 기능설명
---

## Update #2
#### 2018 11 21
- 문제점발견으로 인해 코드 초기화 후 새로운 구조 구축
---

## Update #1
#### 2018 11 18
- 파일 구조 파악 Header 를 포함한 기본틀 제작 시작
---

## START
#### 2018 11 16
- 프로젝트 시작

src/Calendarien/Calendarien.js
import React, { Component } from "react";
import PropTypes from "prop-types";

import Layout from "./common/Layout";
import Header from "./common/Header";
import Footer from "./common/Footer";
import Input from "./common/Input";

import Viewer from "./calendar/Viewer";
import Status from "./calendar/Status";

import { MILLISECONDS, CREATE } from "./utils/dates";
import { format } from "./utils/format";
import { joinstr, isValid } from "./utils/util";

class Calendarien extends Component {

    static defaultProps = {
        style : {
            width: '100%',
            height: 'inherit',
            color: '#262626'
        },
        format: '',
        setDate: '',
        theme: '',
        mode : 'default',
        inputClass: '',
        visibleMyDate: false,
        visibleToday: false,
        disabled : false,
        allowRange : false,
        setRange : [],
        layoutOption: [],
        customizeIcon: [],
        getValue: () => {},
    }

    state = {
        value : MILLISECONDS(),
        selectedValue : 0,
        mode : {
            input : ''
        },
        toggle: {
            input: false
        }
    }

    initialize = () => {
        const { setDate } = this.props;
        if (setDate) {
            this.setState({
                value: CREATE(setDate).getTime()
            })
        }
    }

    handleSelect = (selectedValue) => {
        this.setState({ selectedValue })
    }

    handleMonthChange = (value) => {
        this.setState({ value })
    }

    onMyDate = () => {
        this.initialize();
    }

    onToday = ({value, selectedValue}) => {
        const { format: fm, mode } = this.props;

        this.setState({ value, selectedValue });

        if ( mode === 'input' ) {
            const result = !fm ? format('yyyy-mm-dd') : format(fm, value);
            const { setInputValue } = this;
            setInputValue(result);
        }
    }

    setPropsValue = (selectValue) => {
        const { getValue, format: fm, mode } = this.props;
        const result = !fm ? selectValue : format(fm, selectValue);

        if ( mode === 'input' ) {
            const { setInputValue } = this;
            setInputValue(result);
        }

        getValue(result);
    }

    setInputValue = (value) => {
        this.setState({ mode : { input : value }})
    }

    handleModal = (bool) => {
        this.setState((prevState) => ({ toggle : { ...prevState.toggle, input: bool } }));
    }

    handleModalClear = () => {
        const { getValue } = this.props;
        this.setState({ mode : {input: ''}});
        getValue('');
    }

    handleModalSubmit = () => {
        this.setState({ toggle : {input: false}})
    }

    componentDidMount () {
        this.initialize();
    }

    componentDidUpdate (prevProps, prevState) {
        const { setDate } = this.props;

        if (prevProps.setDate !== this.props.setDate) {
            const is = isValid(setDate);

            if (is) {
                this.initialize();
            } else {
                console.warn('Please fill in the correct date format \n 18 01 01 (x) \n 2018 01 01 (o)');
            }
        }
    }

    render () {

        const {
            handleSelect,
            handleMonthChange,
            setPropsValue,
            onMyDate,
            onToday,
            handleModal,
            handleModalClear,
            handleModalSubmit
        } = this;

        const { 
            style,
            theme,
            inputClass,
            layoutOption,
            customizeIcon,
            visibleMyDate,
            visibleToday,
            disabled,
            setDate,
            mode,
        } = this.props;

        const { 
            value,
            selectedValue,
            toggle,
            mode : _mode
        } = this.state;

        const now = disabled ? Date.now() : 0;
        const classnameJoin = joinstr("calendarien", layoutOption);
        const _toggle = toggle[mode] ? 'toggle__' + mode : ''; 

        return (
            <div className={`demnodey ${_toggle}`} >
                <Input className={inputClass} mode={mode} value={_mode.input} onFocus={handleModal} readOnly />
                <Layout className={classnameJoin} theme={theme} mode={mode} style={style}>

                    { mode !== 'default' && <Status value={_mode.input}/> }

                    <Header 
                        value={value} 
                        customizeIcon={customizeIcon}
                        handleMonthChange={handleMonthChange}
                    />

                    <Viewer
                        value={value}
                        handleSelect={handleSelect}
                        handleMonthChange={handleMonthChange}
                        selectedValue={selectedValue}
                        setPropsValue={setPropsValue}
                        disabled={disabled}
                        now={now}
                    />

                    <Footer
                        visibleMyDate={visibleMyDate}
                        visibleToday={visibleToday}
                        setDate={setDate}
                        mode={mode}
                        toggle={toggle}
                        onMyDate={onMyDate}
                        onToday={onToday}
                        onClear={handleModalClear}
                        onSubmit={handleModalSubmit}
                    />

                </Layout>
            </div>
        )
    }
}

Calendarien.propTypes = {
    style: PropTypes.shape({ 
        background: PropTypes.string
    }),
    mode: PropTypes.string,
    theme: PropTypes.string,
    setDate: PropTypes.string,
    format: PropTypes.string,
    inputClass: PropTypes.string,
    disabled: PropTypes.bool,
    visibleToday: PropTypes.bool,
    visibleMyDate: PropTypes.bool,
    allowRange: PropTypes.bool,
    setRange: PropTypes.arrayOf(PropTypes.number),
    layoutOption: PropTypes.arrayOf(PropTypes.string),
    customizeIcon: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.object, 
            PropTypes.string,
        ])
    ),
    getValue: PropTypes.func,
}

export default Calendarien;